### ISSUE TITLE
Unable parse runtime variables in Apache 2.2

### ISSUE THREAD

**Issue by justfortherec** [letsencrypt/letsencrypt#1208]
39c83d17d7adb790ec68658a957481c73f63e7e0 adds the `-t` flag to `apache2ctl -t -D DUMP_RUN_CFG` in order to avoid erroneous detection of misconfiguration.

`letsencrypt` does not complain about misconfiguration anymore but fails when trying to parse output of `apache2ctl`:

```
The apache plugin is not working; there may be problems with your existing configuration.
The error was: PluginError('Unable to parse runtime variables',)
```

No wonder, the output is basically empty:
```
# apache2ctl -t -D DUMP_RUN_CFG
Syntax OK
```

Adding the `-t` flag seems but a part of the solution.

-------
There are two discussions of this issue on the community board:
https://community.letsencrypt.org/t/how-is-my-apache-misconfigured-exactly/2326
https://community.letsencrypt.org/t/problems-using-letsencrypt-client-in-beta/1925

**Comment by pde** [comment]
flagging this for @neffs

**Comment by neffs** [comment]
I didn't have time to fully solve it, https://github.com/letsencrypt/letsencrypt/commit/39c83d17d7adb790ec68658a957481c73f63e7e0 only fixed standalone.

Is DUMP_RUN_CFG really needed? Which values are needed from that?

**Comment by kuba** [comment]
#1253 might have some more debugging info

**Comment by justfortherec** [comment]
What variables are stored here exactly? What are they used for?

Output of `apache2ctl` is used in `letsencrypt/letsencrypt-apache/letsencrypt_apache/parser.py`

```python
        stdout = self._get_runtime_cfg(ctl)

        variables = dict()
        matches = re.compile(r"Define: ([^ \n]*)").findall(stdout)
        try:
            matches.remove("DUMP_RUN_CFG")
        except ValueError:
            raise errors.PluginError("Unable to parse runtime variables")

        for match in matches:
            # ...
            parts = match.partition("=")
            variables[parts[0]] = parts[2]
```

So the pattern to match variables is: `Define: <variable name>=<variable value>`

With a fresh installation of Apache 2.4 (on Ubuntu Wily), the following output is generated:
```
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.6. Set the 'ServerName' directive globally to suppress this message
ServerRoot: "/etc/apache2"
Main DocumentRoot: "/var/www"
Main ErrorLog: "/var/log/apache2/error.log"
Mutex default: dir="/var/lock/apache2" mechanism=fcntl 
Mutex watchdog-callback: using_defaults
PidFile: "/var/run/apache2/apache2.pid"
Define: DUMP_RUN_CFG
User: name="www-data" id=33
Group: name="www-data" id=33
```

As far as I can tell, nothing matches the pattern (in this vanilla context). I also have a hard time finding usage of these variables in `letsencrypt_apache`. It would be much easier to find a solution for this if we knew the purpose of these variables.

**Comment by neffs** [comment]
Maybe @jdkasten knows what it is used for?

Deactivating it unfortunately doesn't solve the Apache 2.2 problems. Now I get a dvsni error.

**Comment by jdkasten** [comment]
Sorry, I have been a bit absent here. I am in the process of moving.

I am currently away from my computer, so I can't look at code, but you can
define variables in Apache configs.
IIRC
Define ssl_port 443
Define config_path /etc/path

Listen $ssl_port
Include $config_path

You can also do things like...
Define new_functionality
...
<IfDefine new_functionality>

At one point I had implemented it through a directive search. Apache only
makes one pass at what is enabled though so if you look at configs in the
wrong order... your result might not be == to Apache's.  Think "Include
/conf/*/"

So I had elected to simply rewrite the code to take the variables directly
from the executable.

Hope this helps. Sorry about the problems.

**Comment by jdkasten** [comment]
A soft fail might be a reasonable solution for the meantime.

**Comment by justfortherec** [comment]
Apache 2.2 does not have a `Define` directive ([at least not documented](http://httpd.apache.org/docs/2.2/mod/core.html)). The only way of defining these variables in version 2.2 seems to be the `-D` command line option.

Would it be sufficient to parse `/proc/<pid>/cmdline` for them?

In any case, a soft fail seems to be a sensible solution for the time being.

**Comment by justfortherec** [comment]
Can I help in any further way to get this fixed?

**Comment by eXpl0it3r** [comment]
I'm experiencing the same issue (Debian, Apache 2.2). For now I just used the [webroot plugin](https://letsencrypt.readthedocs.org/en/latest/using.html#plugins) instead.

**Comment by justfortherec** [comment]
@pde, does the "Nice for 1.0" tag mean that Apache 2.2 is not officially supported?

As I understand it, this tag means a feature is optional and "nice to have" instead of critical for proper functionality. However, without this issue  being fixed lets-encrypt won't work with Apache 2.2. In light of version 2.2 being the most used version of Apach2 (cf. [1], [2], and [3]), this is a real showstopper for lets-encrypt.

I highly recommend reassessing the severity of this issue.

[1]: http://w3techs.com/technologies/details/ws-apache/2/all
[2]: http://trends.builtwith.com/Web-Server/Apache-2.4
[3]: http://trends.builtwith.com/Web-Server/Apache-2.2

**Comment by xezpeleta** [comment]
Same here.

I have mod_jk mod enabled in my Apache 2.2 (Debian Wheezy, LTS) and one of my vhosts is configured with "JkMount" directive. 

**letsencrypt.log:**
```python
2015-11-10 10:14:30,234:DEBUG:letsencrypt.cli:Root logging level set at 30
2015-11-10 10:14:30,235:INFO:letsencrypt.cli:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2015-11-10 10:14:30,235:DEBUG:letsencrypt.cli:letsencrypt version: 0.0.0.dev20151108
2015-11-10 10:14:30,235:DEBUG:letsencrypt.cli:Arguments: ['--agree-dev-preview', '--server', 'https://acme-v01.api.letsencrypt.org/directory', '-d', 'mysecretdomain.eus', '--apache']
2015-11-10 10:14:30,235:DEBUG:letsencrypt.cli:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)
2015-11-10 10:14:30,241:DEBUG:letsencrypt.cli:Requested authenticator apache and installer apache
2015-11-10 10:14:30,396:DEBUG:letsencrypt.plugins.disco:Other error:(PluginEntryPoint#apache): Unable to parse runtime variables
Traceback (most recent call last):
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt/plugins/disco.py", line 103, in prepare
    self._initialized.prepare()
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/configurator.py", line 150, in prepare
    self.aug, self.conf("server-root"), self.conf("ctl"))
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/parser.py", line 39, in __init__
    self.update_runtime_variables(ctl)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt_apache/parser.py", line 104, in update_runtime_variables
    raise errors.PluginError("Unable to parse runtime variables")
PluginError: Unable to parse runtime variables
2015-11-10 10:14:30,397:DEBUG:letsencrypt.display.ops:No candidate plugin
2015-11-10 10:14:30,397:DEBUG:letsencrypt.cli:Selected authenticator None and installer None
```

It works well using "certonly", but I'd like to configure the automatic renewal for all the domains. I understand that we need to use the "run" mode for that...

**Comment by Adzter** [comment]
Also having the same problem, it seems that when
`apache2ctl -t -D DUMP_RUN_CFG`

is ran it leaves an Apache process running in the background

**Comment by pde** [comment]
Sorry I missed some notifications on this ticket.  An update on process: this being flagged as **nice for 1.0** means that we haven't prioritised fixing this before the Public Beta launch on the 3rd of December.  However if someone can produce a pull request to fix it, we'd be interested in taking it, and there's a chance that @swartzcr will be able to work on this (reviewing a PR if someone can get started, or writing one if not) in the next week or two.

**Comment by justfortherec** [comment]
I'd be happy to work on it. However it's hard to get started without above questions being answered:

> Is DUMP_RUN_CFG really needed?
> Which values are needed from that?
> What variables are stored here exactly?
> What are they used for?
> Would it be sufficient to parse `/proc/<pid>/cmdline` for them?

**Comment by jdkasten** [comment]
I believe I answered this above, but.... These are variables specified
within the Apache config. You can experiment or read documentation.

http://httpd.apache.org/docs/2.4/mod/core.html#define

They are used throughout configs. ${variable}. and in ifdef statements. You
can't interpret the config accurately if you can't follow these paths.

Most obvious case ```Include ${var}```

Feel free to make it gracefully fail...

Configurator.version can be of use.

Change parser.get_aug (I think it is what it is called) to use .get instead
of dict[key].

I apologize for the brevity, but I am answering afk.

**Comment by justfortherec** [comment]
You're linking documentation of Apache 2.4. As previously said: This *directive does not exist in Apache 2.2* which this entire bug is about.

Can you please have a look at my questions in light of this information (especially the first and the last question)? Thanks

**Comment by SwartzCr** [comment]
So the code seems to be looking for things in the output of the apache2ctl -t -D DUMP_RUN_CFG
of the form: Define: x = y
On my machine at least there is nothing of that form in the output (and similarly mentioned in https://github.com/letsencrypt/letsencrypt/issues/1208#issuecomment-152827754)
this results in a parser.variables dictionary of the form {}
The exception that's raised is if it can't find DUMP_RUN_CFG in the output.
The only two functions these variables get used in are get_arg and _exclude_dirs in parser.py
It seems to me that we could just return if DUMP_RUN_CFG isn't returned and carry on with the auth, as that array seems to be empty all the time anyway.
On the other hand I'm not sure what happens if there are variables specified in the return value of apache2ctl -t -D DUMP_RUN_CFG and they don't get proliferated.
Does anyone have an example of an output that will result in a non-empty variables dict?

**Comment by pde** [comment]
Yes, I see that output if you have something like `Define VARIABLE 3` in the running configuration file.  See [the docs](https://people.apache.org/~rjung/mod_define/mod_define.html)

**Comment by justfortherec** [comment]
@pde, is that with Apache 2.2 or 2.4?

**Comment by pde** [comment]
@justfortherec you're right, I thought that system had apache2.2 but it was 2.4.  My current read is that if an Apache 2.2 system supports `Define` it's only because of `mod_define`, which isn't packaged on Debian-based systems.

**Comment by pde** [comment]
So what @jdkasten means by "fail gracefully" is that if you see `Define` directives, and `mod_define` enabled on an Apache 2.2 system, just report an error saying that A2.2 + mod_define isn't supported.

**Comment by SwartzCr** [comment]
The issue as I understand it is that the apache2ctl command we're running doesn't provide the same output in apache 2.2 as it does in 2.4 meaning that for a 2.2 system we won't see these Define statements.  I've been looking through apache documentation (without much success) to try to find how to get apache2ctl to spit these out on 2.2

The code as it's written now throws an error if Define: DUMP_RUN_CFG isn't output, thinking that the apache2ctl call has failed. Since 2.2 doesn't output anything except "Syntax OK" the client errors out. So we could for example not error out in this case and just tell 2.2 users that their mod_define statements aren't supported, though I'm not sure what effect this has. Presumably we could also make this version dependent so it wouldn't effect the case for v2.4 users.

**Comment by pde** [comment]
@SwartzCr the most cautious solution would be to error out on 2.2 if a Define directive is present (that should be a fairly rare occurrence, especially since Debian-based OSes never seem to have shipped mod_define for Apache 2.2).

By default my instinct would be to go with the cautious approach, but if you think you understand and can control for possible breakage via `mod_define`, we could try to function in its presence.

**Comment by chaitanyanettem** [comment]
Is there a workaround for this issue?

**Comment by SwartzCr** [comment]
peter just posted a workaround that will allow Let's Encrypt to continue on Apache 2.2 systems, but not include any mod_define contributions.

I'm working on a fix for this that will add an error if there are mod_define statements that are not getting included in apache2.2 (by default there shouldn't be any) but I'm running into a few problems.

The first place you might put this check is in the call update_runtime_variables in parser.py on line 104 - instead of erroring out, you would check to see if there are any mod_define statements via augeas and error only if there are. The problem with this is that no files have yet been loaded by augeas to check against.

Alternately, one could check before this gets tested, unfortunately the parser.variables object gets read every time get_arg is called, which also doesn't seem like the right place to be checking this

**Comment by SwartzCr** [comment]
additionally, there's some circular dependency shenanigans going on - since init_modules calls get_arg which looks at the variables list, but is called before (and may be part of) the files get parsed by augeas. I'm going to try to look to see what exactly is happening in these calls, but I'm at a point with the code where I'm not really sure what it's doing

**Comment by mquandalle** [comment]
I’m also encountering this error on Apache/2.2.22 (Ubuntu), is there any (temporary) way to bypass the failing test and continue the installation process?

**Comment by ehsaan** [comment]
I had the same issue, apparently my apache setup did not report as letsencrypt parser expected (it just says Syntax OK).
To go forward, I did a "certonly" and "--manual". (./letsencrypt-auto certonly -d domain.net ... --manual) It means I had to create files for verification manually, and then setup Apache config using PEM files it created.

**Comment by awekas** [comment]
Have the same error. Have also post my logs to letsencrypt help forum: https://community.letsencrypt.org/t/server-certificate-error/6180/7

**Comment by plutext** [comment]
I imagine heaps of people get this same error, but only a small proportion of them find this issue, and a smaller portion again post a comment :-)

**Comment by SwartzCr** [comment]
I am in fact working on this. Bypassing the failing test uncovers other errors that I am trying to debug. Currently it seems that augeas is unable to find VirtualHost directives on apache 2.2 which I am looking into

**Comment by cjd118** [comment]
Same issue on Apache 2.2 (Debian). Ended up using webroot plugin instead and manually adding the certificate info to vhosts.

**Comment by pde** [comment]
Closed by #1961.

**Comment by pde** [comment]
Due to some undiagnosed bugs, we're not actually going to enable Apache 2.2 support until 0.2.1.
